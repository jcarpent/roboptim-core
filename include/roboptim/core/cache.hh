// Copyright (C) 2014 by Benjamin Chr√©tien, CNRS-LIRMM.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_CACHE_HH
# define ROBOPTIM_CORE_CACHE_HH

# include <list>
# include <vector>
# include <boost/unordered_map.hpp>

namespace roboptim
{
  /// \brief LRU (Least Recently Used) cache.
  /// \tparam K type for keys.
  /// \tparam V type for values.
  /// Some of the ideas used here come from
  /// Tim Day's "LRU cache implementation in C++"
  /// (http://timday.bitbucket.org/lru.html)
  template <typename K, typename V>
  class LRUCache
  {
  public:
    /// \brief Type of keys.
    typedef K key_t;

    /// \brief Type of values.
    typedef V value_t;

    typedef std::list<key_t> keyTracker_t;
    typedef std::vector<value_t> valuePool_t;

    /// \brief Map from key to iterator in the value pool.
    typedef boost::unordered_map
    <key_t, typename valuePool_t::iterator> map_t;

    typedef typename map_t::const_iterator const_iterator;
    typedef typename map_t::iterator       iterator;

  public:
    /// \brief Constructor.
    /// Note: all the memory is allocated in the constructor.
    /// \param size maximum size of the cache.
    LRUCache (size_t size = 10);

    /// \brief Destructor.
    ~LRUCache ();

    /// \brief Size of the cache.
    size_t size () const;

    /// \brief Change the size of the cache.
    void resize (size_t size);

    /// \brief Find an element in the cache.
    const_iterator find (const K& key) const;

    /// \brief Iterator to the beginning of the cache.
    iterator begin ();

    /// \brief Iterator to the end of the cache.
    iterator end ();

    /// \brief Iterator to the beginning of the cache.
    const_iterator cbegin () const;

    /// \brief Iterator to the end of the cache.
    const_iterator cend () const;

    /// \brief Access a cached element.
    /// \param key key to the element.
    /// \return reference to the element.
    V& operator [] (const K& key);

    /// \brief Insert a value into the cache.
    /// \param key key of the element.
    /// \param value value of the element.
    void insert (const K& key, const V& value);

    /// \brief Clear the cache.
    void clear ();

    /// \brief Display the cache on the specified output stream.
    virtual std::ostream& print (std::ostream&) const;

  private:
    /// \brief Allocate memory based on the cache's size.
    void allocate ();

    /// \brief Update an existing cached value.
    void update (iterator iter, const V& value);

    /// \brief Update an existing cached value.
    V& update (iterator iter);

    /// \brief Insert a value into the cache.
    V& insert (const K& key);

    /// \brief Notice the tracker that the key was used.
    void bump (const K& key);

    /// \brief Find Least Recently Used element.
    const_iterator findLRU () const;

  private:
    /// \brief Size of the cache.
    size_t size_;

    /// \brief Tracker for the least recently used elements.
    keyTracker_t tracker_;

    /// \brief Map containing iterators to cached values.
    map_t map_;

    /// \brief Pool of values to prevent reallocations.
    valuePool_t pool_;
  };

  template <typename K, typename V>
  std::ostream&
  operator<< (std::ostream& o, const LRUCache<K,V>& cache);

} // end of namespace roboptim

# include <roboptim/core/cache.hxx>

#endif //! ROBOPTIM_CORE_CACHE_HH
