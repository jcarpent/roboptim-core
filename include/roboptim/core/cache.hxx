// Copyright (C) 2014 by Benjamin Chr√©tien, CNRS-LIRMM.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_CACHE_HXX
# define ROBOPTIM_CORE_CACHE_HXX

# include <algorithm>

namespace roboptim
{
  template <typename K, typename V>
  LRUCache<K,V>::LRUCache (size_t size)
    : size_ (size),
      tracker_ (),
      map_ (),
      pool_ (size)
  {
  }

  template <typename K, typename V>
  LRUCache<K,V>::~LRUCache ()
  {}

  template <typename K, typename V>
  size_t LRUCache<K,V>::size () const
  {
    return size_;
  }

  template <typename K, typename V>
  void LRUCache<K,V>::resize (size_t size)
  {
    if (size != size_)
      {
	size_ = size;
	allocate ();
      }
  }

  template <typename K, typename V>
  void LRUCache<K,V>::allocate ()
  {
    // FIXME: breaks iterators in map_?
    map_.rehash (size_);
    tracker_.resize (size_);
  }

  template <typename K, typename V>
  void LRUCache<K,V>::clear ()
  {
    map_.clear ();
    tracker_.clear ();
  }

  template <typename K, typename V>
  void LRUCache<K,V>::insert (const K& key, const V& value)
  {
    V& v = insert (key);
    v = value;
  }

  template <typename K, typename V>
  V& LRUCache<K,V>::insert (const K& key)
  {
    typename valuePool_t::iterator v_it;

    // If the key is already in the map
    iterator iter = map_.find (key);
    if (iter != map_.end ())
      {
	bump (iter->first);
	return *(iter->second);
      }

    // If the cache is full, remove the LRU element
    if (map_.size () >= size_)
      {
	const_iterator it = findLRU ();

	// Remove the key from the tracker
	tracker_.pop_front ();

	// Update the iterator for the pool
	v_it = it->second;

	// Remove the LRU element from the map
	map_.erase (it);
      }
    else // cache not full
      {
	// Get the first free pool element
	v_it = pool_.begin ()
	  + static_cast<typename valuePool_t::iterator::difference_type>
	  (tracker_.size ());
      }

    // Add the new key to the tracker
    tracker_.push_back (key);

    // Add the new key to the map
    typename map_t::value_type p (key, v_it);
    map_.insert (p);

    return *v_it;
  }

  template <typename K, typename V>
  void LRUCache<K,V>::update (iterator iter, const V& value)
  {
    bump (iter->first);
    *(iter->second) = value;
  }

  template <typename K, typename V>
  V& LRUCache<K,V>::update (iterator iter)
  {
    bump (iter->first);
    return *(iter->second);
  }

  template <typename K, typename V>
  typename LRUCache<K,V>::iterator
  LRUCache<K,V>::begin ()
  {
    return map_.begin ();
  }

  template <typename K, typename V>
  typename LRUCache<K,V>::iterator
  LRUCache<K,V>::end ()
  {
    return map_.end ();
  }

  template <typename K, typename V>
  typename LRUCache<K,V>::const_iterator
  LRUCache<K,V>::cbegin () const
  {
    return map_.cbegin ();
  }

  template <typename K, typename V>
  typename LRUCache<K,V>::const_iterator
  LRUCache<K,V>::cend () const
  {
    return map_.cend ();
  }

  template <typename K, typename V>
  typename LRUCache<K,V>::const_iterator
  LRUCache<K,V>::find (const K& key) const
  {
    return map_.find (key);
  }

  template <typename K, typename V>
  V& LRUCache<K,V>::operator [] (const K& key)
  {
    typename map_t::iterator it = map_.find (key);

    if (it != map_.end ())
      return update (it);
    else
      return insert (key);
  }

  template <typename K, typename V>
  typename LRUCache<K,V>::const_iterator
  LRUCache<K,V>::findLRU () const
  {
    // Find the LRU element
    return map_.find (tracker_.front ());
  }

  template <typename K, typename V>
  void LRUCache<K,V>::bump (const K& key)
  {
    // TODO: optimize this (remove the need for a search algorithm)
    typename keyTracker_t::iterator
      iter = std::find (tracker_.begin (), tracker_.end (), key);
    assert (iter != tracker_.end ());

    tracker_.splice (tracker_.end (),
                     tracker_,
                     iter);

    assert (tracker_.size () <= size_);
  }


  template <typename K, typename V>
  std::ostream& LRUCache<K,V>::print (std::ostream& o) const
  {
    o << "{";
    for (const_iterator iter = map_.begin ();
         iter != map_.end (); iter++)
      {
	if (iter != map_.begin ())
	  o << ", ";
	o << iter->first << ": " << *(iter->second);
      }
    o << "}";

    return o;
  }

  template <typename K, typename V>
  std::ostream&
  operator<< (std::ostream& o, const LRUCache<K,V>& cache)
  {
    return cache.print (o);
  }
} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_CACHE_HXX
