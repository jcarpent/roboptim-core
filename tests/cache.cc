// Copyright (C) 2014 by Benjamin Chretien, CNRS-LIRMM.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <iostream>

#include <roboptim/core/io.hh>
#include <roboptim/core/cache.hh>

using namespace roboptim;

// Output stream
boost::shared_ptr<boost::test_tools::output_test_stream> output;

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE (cache)
{
  boost::shared_ptr<boost::test_tools::output_test_stream>
    output = retrievePattern ("cache");

  size_t cache_size = 5;
  LRUCache<std::string, float> cache (cache_size);

  (*output) << cache << std::endl;

  cache.insert ("x₀", 42.);
  (*output) << cache << std::endl;
  cache.insert ("x₁", 1.);
  cache.insert ("x₀", 0.);
  cache.insert ("x₂", 20.);
  cache.insert ("x₂", 2.);
  cache.insert ("x₃", 3.);
  cache.insert ("x₄", 4.);

  // Note: we do not print when the cache has multiple
  // elements since the print order is not deterministic,
  // thus it cannot be used for tests.

  // x₅ overwrites x₁'s cached value (LRU)
  cache.insert ("x₅", 5.);

  (*output) << "x₀ = " << cache["x₀"] << std::endl;
  (*output) << "x₂ = " << cache["x₂"] << std::endl;
  (*output) << "x₃ = " << cache["x₃"] << std::endl;
  (*output) << "x₄ = " << cache["x₄"] << std::endl;
  (*output) << "x₅ = " << cache["x₅"] << std::endl;

  // /!\ x₁ no longer in cache, so a new element is added,
  // and one removed (x₀)
  (*output) << "x₁ = " << cache["x₁"] << std::endl;

  cache.clear ();
  (*output) << cache << std::endl;

  cache.insert ("x₀", 37.);
  (*output) << cache << std::endl;

  // Test for heap memory
  // #1: fixed size matrices
  typedef Eigen::Matrix<double, 100, 100> matrix_t;
  LRUCache<int,matrix_t> matrix_cache (cache_size);

  for (int i = 0; i < 500; ++i)
  {
    matrix_cache.insert (i, matrix_t::Zero ());
    BOOST_CHECK (matrix_cache[i].isZero ());
  }

  // #2: dynamic size matrices
  typedef Eigen::MatrixXd dynMatrix_t;
  LRUCache<int,dynMatrix_t> dynMatrix_cache (cache_size);

  for (int i = 0; i < 500; ++i)
  {
    dynMatrix_cache.insert (i, dynMatrix_t::Zero (100, 100));
    BOOST_CHECK (dynMatrix_cache[i].isZero ());
  }

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()

