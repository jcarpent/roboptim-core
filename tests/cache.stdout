{}
{x₀: 42}
x₀ = 0
x₂ = 2
x₃ = 3
x₄ = 4
x₅ = 5
x₁ = 0
{}
{x₀: 37}
